### ISSUE TITLE
.Net Core for Microsoft.Graph.Communications.Calls.Media 

### ISSUE THREAD

**Issue by stefan2410** [microsoftgraph/microsoft-graph-comms-samples#193]
We build our applications in .Net Core 3.1 .
 How to reference the  Microsoft.Graph.Communications.Calls.Media  and Microsoft.Skype.Bots.Media ?

**Comment by jsweiler** [comment]
I also am interested in this. I tried running a .Net Core 3.1 app with these dlls but I get errors when trying to initialize the media platform. The errors I get are errors loading various system dlls. 

Is there a way to use these in a .net core 3.1 app?

**Comment by ssulzer** [comment]
@stefan2410 @jsweiler 
Hi folks,
An app-hosted real-time media bot (using the Microsoft.Skype.Bots.Media API) **must** use the classic .NET Framework; .NET core is not supported. We cannot offer a time frame right now of if/when that might change.
https://docs.microsoft.com/en-us/microsoftteams/platform/bots/calls-and-meetings/requirements-considerations-application-hosted-media-bots

**Comment by stefan2410** [comment]
@ssulzer The problem is that MS strongly recommends to develop new code in .Net 3.1 (LTS) and the architecture of .Net Core applications is different and mostly the performance is superb for back-end solutions such as the Bots we are talking.
How to build  a new product with 5 years old code, when the developers we learn and work only in .Net Core ?

**Comment by ssulzer** [comment]
Hi @stefan2410 
The issue concerns more than the System.Threading.Tasks.Dataflow library; the Microsoft.Skype.Bots.Media library has other dependencies that .NET core does not provide. There is some work in-progress now that will help enable us to support .NET core/.NET 5 in the future. Of course we want to support .NET core, but it will take some time to get there. Thanks.

**Comment by stefan2410** [comment]
@ssulzer Just one thought. The .Net Core applications can run only  in Windows server.
Like the  WPF and Winforms in .Net Core 3.1. they run only in windows

**Comment by JapNolt** [comment]
I am fine if there would still be a dependency on Windows but it would be nice if our projects could target .Net Core. It's good to hear that there's work being done on this.

**Comment by AlexSapple** [comment]
I'm also interested - Just want to add support to this! 

thanks everyone.

**Comment by neomedved** [comment]
Any updates on this after .Net 5 release?

**Comment by prdx2000** [comment]
How long is it going to take to have this working on .Net 5?

**Comment by AHComp** [comment]
The ticket has been open for almost a year now and nothing is happening point I find it very embarrassing for Microsoft that everything is lagging behind in important parts such as teams and in communication such as Sip, and it's a big problem to deploy the ancient code in Azure because in Azure's new resource manager concept, the old projects no longer run and could not deployed. when do you deliver a new, up-to-date code that runs on the current basis specified by Microsoft ?

**Comment by kieronlanning** [comment]
@AHComp I've got an end to end working using the Cloud Service (extended support) resource that is ARM deployable, while it's not .NET Core/ cross-platform it does make it easier to manage.

**Comment by stefan2410** [comment]
@kieronlanning  It can't be cross-platform in any case because of th Skype.Media, which requires Windows Server and Media Foundation.   But the update in .Net 5  is very important because we need to use modern libraries and features. 
@ssulzer  is there a roadmap?

**Comment by kieronlanning** [comment]
@stefan2410 Yeah, that requirement has been painful. We did investigate Windows Containers too, but they're not supported either due to the hardware access requirements.

We have updated to .NET Framework 4.8 (which supports .NET Standard 2.0 - .NET 5 is Standard 2.1), but that's the limit. While we use the .NET 5 Tool chain for building, we can't progress higher.

**Comment by ssulzer** [comment]
No concrete roadmap or ETA to share yet regarding Microsoft.Skype.Bots.Media support for .NET core/.NET 5 (but work is in-progress on this).

**Comment by ankitbko** [comment]
Really hoping that support for .NET 5 comes soon. Meanwhile I have created a *StarterBot* on `ASP.NET Core 2.1` project targeting `net48`. It is mostly based on `AudioVideoPlaybackBot` but *without* any of its functionality. All it does is log message for every video, audio and screen share events. Its meant to a template to get started with creating new bot that needs these functionality. This sample works exactly like any other asp.net core project so you can host it on Kestrel. The other samples were either on Service Fabric or Work roles which are both quite aged technology. Its also quite easy to debug it locally (no more Azure Compute emulator). Once I have documented it I will raise a PR here. https://github.com/ankitbko/microsoft-graph-comms-samples/tree/aspnetcore/Samples/V1.0Samples/StarterBotSamples

**Comment by stefan2410** [comment]
@ankitbko, as a user also ,I would say that your effort is very nice, but  we can  "debug it locally (no more Azure Compute emulator)",  as it is now,  making the WorkerRole startup project (console application in.Net Framework). with minor changes in AzureConfiguration (another class LocalConfiguration?)
Of course,  "Fabric or Work roles which are both quite aged technology"  are Required in production. Otherwise you would need your load balancing service in case that you have many VM instances.

**Comment by ankitbko** [comment]
@stefan2410 Yup you can debug the WorkerRole project locally as well. When I first started learning about Media Calling bot couple of days ago I felt like I am back in 2015 😆. It has been ages since I had to touch Worker Role project. I didn't see any benefit of sticking to it. I am too used to ASP.NET Core. 
Also they are not really "required" for production. There is already an example with AKS if you want a scalable architecture. I haven't explored if the bot will work in App Service.

**Comment by stefan2410** [comment]
@ankitbko AKS can support Windows Servers?  because if you want to play anything with Media Platform, you need windows and media.Foundation.
AppService does not work definitely.
You said before thta you  can debug it without  emulator. I simply mentioned that this is not an issue of ASP Net Core.  We can do it even now. There are other serious issues that we need the .Net 5.

**Comment by ankitbko** [comment]
@stefan2410 Yes AKS [supports Windows nodes](https://docs.microsoft.com/en-us/azure/aks/windows-container-cli). You can find the sample here https://github.com/microsoftgraph/microsoft-graph-comms-samples/tree/master/Samples/V1.0Samples/AksSamples/teams-recording-bot

The doc folder in that sample has learnings and some good documentation for this.

**Comment by mosoftwareenterprises** [comment]
> No concrete roadmap or ETA to share yet regarding Microsoft.Skype.Bots.Media support for .NET core/.NET 5 (but work is in-progress on this).

Hi @ssulzer any update on this comment from from nearly a year ago?

**Comment by ssulzer** [comment]
@mosoftwareenterprises 
Sorry, no useful update to share. The SDK port to .NET Core is still in-progress. I _hope_ we can ship it by end of this year.

**Comment by Mishra-Saurabh** [comment]
> 

If we are not able to use skype.bot .media in .net core then is there any other way to use in place of skype. bot. media in.net core.

**Comment by decker405** [comment]
@ssulzer any updates on the "end of this year" timeframe?

**Comment by ssulzer** [comment]
@decker405 
It won't be end of this year, but we are getting closer. Right now, our current ETA is April/May timeframe. We are targeting .NET6.

**Comment by decker405** [comment]
Appreciate the update @ssulzer

**Comment by martenekblad** [comment]
Thanks @ssulzer !
Very appriated utdate!
Please prioritize this since it has been sinking us for a while now.

**Comment by roberto-gianisella** [comment]
Hi @ssulzer, is the April/May ETA still happening or do you expect delays?

**Comment by ssulzer** [comment]
@roberto-gianisella @martenekblad 
Currently we are still on track for an April or May release. I'll update this thread as we get closer. Thanks!

**Comment by desek** [comment]
May is coming fast. Any updates to planned release date?

**Comment by AHComp** [comment]
Earth to Microsoft.  mayday the flight from Mars to Earth will take longer than your migration.  It's been almost 3 years now and now release date ?

**Comment by ssulzer** [comment]
The ship will be rolling out on the launch pad soon, but likely will be a June lift-off.

The two SDK packages, Microsoft.Skype.Bots.Media and Microsoft.Graph.Communications.Calls.Media are on somewhat different schedules. We will release a new Microsoft.Skype.Bots.Media library with .NET6 a bit ahead of the Microsoft.Graph.Comms.Calls.Media library update (since one is a dependency of the other).

The goal of this initial release is to help unblock (finally) development of bot applications on .NET6+.
But we do not yet have enough runtime experience with our own Teams/Skype real-time media bots on .NET core, so we advise the community to be cautious migrating production media bots onto .NET core. (i.e., please test thoroughly!)

We are going to make the .NET6 SDK libraries available ahead of any corresponding updates to the public samples. 

cc: @rahulva-msft  @zhengni-msft

**Comment by AHComp** [comment]
Ok I plan my test Migration in June now hope it will be not Oktober or December, the nights are very cold in December on Mars.

**Comment by ssulzer** [comment]
The next Microsoft.Skype.Bots.Media library release is expected next week. It is undergoing release testing. It will have support for .NET6 (but, like I mentioned earlier, consider the .NET6 version it an alpha/preview).
I am checking on the Microsoft.Graph.Communications.Calls.Media library update. Hopefully that will also get out in June.
Thanks.

**Comment by AHComp** [comment]
Hello Stephen, are you okay?
After 3 years waiting, I plan the migration for june and now you say perhaps ? 
Are you kidding me or was it big joke ?

**Comment by stefan2410** [comment]
@AHComp    please   STOP . We are engineers here, not hooligans.

**Comment by AHComp** [comment]
I'm sorry, is it really no longer allowed to criticize?
I wait now for 3 years and now that ?

**Comment by MathieuDiepman** [comment]
@AHComp Not sure what you are trying to achieve with your criticism. I’m sure Steven is aware that it’s been three years and that it’s been delayed. Perhaps offering help (in release testing or review) instead would be more effective.

We now know that release testing is happening. Personally, I appreciated the update and am patiently waiting for the library to be released.

**Comment by ssulzer** [comment]
@stefan2410 @MathieuDiepman 

Folks, the Microsoft.Skype.Bots.Media library version 1.26.0.94 released this week includes .NET6 support.
https://www.nuget.org/packages/Microsoft.Skype.Bots.Media/1.26.0.94-alpha

Notes:
1. We are still waiting on a corresponding update to the Microsoft.Graph.Communications.Calls.Media library; it will be a "-beta" release.
2. The net6 target library has a somewhat different set of runtime dependencies.
3. It is not clear whether running the bot media runtime on .NET6 is compatible with ngrok; I don't think we got that to work yet.

**Comment by ssulzer** [comment]
@stefan2410 @MathieuDiepman 
And folks we now have the Microsoft.Graph.Communications.Calls and .Media libraries now updated with .NET6 support!

[NuGet Gallery | Microsoft.Graph.Communications.Calls 1.2.0.7270](https://www.nuget.org/packages/Microsoft.Graph.Communications.Calls/1.2.0.7270)

[NuGet Gallery | Microsoft.Graph.Communications.Calls.Media 1.2.0.7270](https://www.nuget.org/packages/Microsoft.Graph.Communications.Calls.Media/1.2.0.7270)

**Comment by ssulzer** [comment]
A follow-up note regarding .NET6 support that is likely most relevant when testing a bot locally behind an ngrok reverse-proxy.
If the bot sets MediaPlatformSettings.MediaPlatformInstanceSettings.**InstancePublicIPAddress** to **0.0.0.0** (i.e., IPAddress.Any), the bot media platform may fail with an exception on startup.
If that happens, a workaround is to set InstancePublicIPAddress to an IPv4 address of the local machine.

**Comment by chanheelee-thumb** [comment]
@ssulzer Hi Stephen, I'm not quite as familiar with development in .NET but I've been reading up on the documentation and is there any reason I would be getting this error despite not having any PublicKeyTokens listed on my csproj file? My package reference just looks like this in my csproj file: `<PackageReference Include="Microsoft.Skype.Bots.Media" Version="1.27.0.2-alpha" />`

The error: 
`"Could not load file or assembly 'Microsoft.Skype.Bots.Media, Version=1.27.0.2, Culture=neutral, PublicKeyToken={some_value}'. The system cannot find the file specified.\n"`

For more background information I'm trying to use the Microsoft.Skype.Bots.Media package in a net 6.0 project to record audio from a Teams meeting.

**Comment by adityaramgopal** [comment]
@chanheelee-thumb would you be able to share your csproj file (just the csproj file) and perhaps the *deps.json file in the `obj` folder

**Comment by chanheelee-thumb** [comment]
Hi @adityaramgopal I kept getting an error saying 'X could not be updated/installed' from Nuget package manager on Visual Studio (Mac 17.6.6) so I added it through dotnet commands. This is the csproj code: 
```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <LangVersion>latest</LangVersion>
    <Nullable>enable</Nullable>
    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
    <AnalysisLevel>preview-recommended</AnalysisLevel>
  </PropertyGroup>

  <ItemGroup>
    <None Remove="Helpers\NewFile.txt" />
    <None Remove="Interfaces\NewFile.txt" />
    <None Remove="Models\bot\" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.17.2" />
    <PackageReference Include="Microsoft.CognitiveServices.Speech" Version="1.24.1" />
    <PackageReference Include="Microsoft.Graph" Version="4.45.0" />
    <PackageReference Include="Azure.Identity" Version="1.7.0" />
    <PackageReference Include="Microsoft.Graph.Communications.Calls" Version="1.2.0.7270" />
    <PackageReference Include="Microsoft.Graph.Communications.Calls.Media" Version="1.2.0.7270" />
    <PackageReference Include="Microsoft.Graph.Communications.Core" Version="1.2.0.7270" />
    <PackageReference Include="Microsoft.Skype.Bots.Media" Version="1.27.0.2-alpha" />
    <PackageReference Include="System.Net.Http.Formatting.Extension" Version="5.2.3" />
    <PackageReference Include="AdaptiveCards" Version="2.7.3" />
    <PackageReference Include="AdaptiveCards.Templating" Version="1.3.0" />
    <PackageReference Include="Microsoft.Bot.Builder.Calling" Version="3.20.1" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.WebApiCompatShim" Version="2.2.0" />
  </ItemGroup>

  <ItemGroup>
    <Content Update="appsettings.json">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
  </ItemGroup>
  <ItemGroup>
    <Folder Include="Models\bot\" />
  </ItemGroup>
</Project>
```

**Comment by adityaramgopal** [comment]
@chanheelee-thumb are you trying to run it on a Mac? This library has the limitation that it's Windows-only.

**Comment by chanheelee-thumb** [comment]
@adityaramgopal I saw in the documentation that "Microsoft.Skype.Bots.Media" Version="1.27.0.2-alpha" supports net 6.0. Was I incorrect to assume that net 6.0 compatible == mac and windows compatible?

**Comment by adityaramgopal** [comment]
.net6.0 is supposed to help you build cross platform apps. But the library needs to be able to support cross-platform. We still have many components under the covers that are still tied to Windows. We mentioned that in our Readme that opens up when you install the nuget.

**Comment by chanheelee-thumb** [comment]
My apologies. I appreciate the clarification, thank you!

**Comment by adityaramgopal** [comment]
Eventually we may support cross-platform but at this moment it's unfortunately Windows-only. Let us know if you run into issues on Windows.

**Comment by SydorenkoEgor** [comment]
Hey @stefan2410 @ssulzer 
Could you please help us?

We have a working application on .net Framework 4.7.2 which is using
Next Nuget packages:
Microsoft.Graph.Communication.Calls
Microsoft.Graph.Communication.Calls.Media
Microsoft.Skype.Bots.Media

We noticed that after updating these packages to the newer versions (1.2.0.9486 for the first 2 and 1.27.0.2-alpha for Skype.Bot.Media) we observed the next issue:
Sometimes when the bot joins the call where participants are already present, the participant collection event "Participants.OnUpdated" will fire only once with the representation of the participant of the bot that joined the call in the AddedResources collection (Information about new participants that will try to join after bot will be fired in the event as well). But when we are testing the same behavior on .netFramework versions of packages we observe that
"Participants.OnUpdated" will be fired as many times as we have participants on the call at the moment when the bot joins the call + 1 for the bot.

After this, we enabled logs and observed that this message sometimes in .net6  in the described case will come only with bot information. But in .net framework 4.7.2 such message will be triggered for all users that are already in the call.
`{
  "value": [
    {
      "changeType": "Updated",
      "resourceUrl": "/communications/calls/821f6500-3a6a-4b3d-be8b-5c824f2479b9/participants",
      "resource": "/app/calls/821f6500-3a6a-4b3d-be8b-5c824f2479b9/participants",
      "resourceData": [
        {
          "@odata.type": "#microsoft.graph.deltaParticipants",
          "participants": [
            {
             ///Paticipant info
            }
          ],
          "sequenceNumber": 3
        }
      ]
    }
  ]
}`

Could you please help us maybe we are doing something wrong?

Thank you!

**Comment by vigbav36** [comment]
@adityaramgopal Hey is there an ETA on moving away from the windows dependency ?  Just wanted to know if its on the roadmap for the near future , thanks !

**Comment by ssulzer** [comment]
Hi @vigbav36 
No ETA to share; not in the near future.

**Comment by SharePointX** [comment]
.net6.0 will be out of support in a few months. Are there any plans to support .net8.0?

**Comment by tedlatte** [comment]
If you want an easier way to build cross platform apps, another alternative is to use [Recall.ai ](https://www.recall.ai/?utm_source=zoom_dev_forum&utm_medium=forum_answer&utm_campaign=net-core-3.1&utm_content=Recall.ai_API)for your meeting bots instead. It’s a simple 3rd party API that lets you use meeting bots to get raw audio/video from meetings without you needing to spend months to build, scale and maintain these bots.
Let me know if you have any questions!

**Comment by ssulzer** [comment]
> .net6.0 will be out of support in a few months. Are there any plans to support .net8.0?

Hi @SharePointX 
Yes, we will add a .net8 target in the media nuget by end of this year; however, the current .net6 library does not have any known issues running on the .net8 runtime. Thanks.

**Comment by lojoh** [comment]
> > Eventually we may support cross-platform but at this moment it's unfortunately Windows-only. Let us know if you run into issues on Windows.
> 
> [@adityaramgopal](https://github.com/adityaramgopal) Hey is there an ETA on moving away from the windows dependency ? Just wanted to know if its on the roadmap for the near future , thanks !

> Hi [@vigbav36](https://github.com/vigbav36) No ETA to share; not in the near future.

Hey. Any updates on this? Is there any plans on moving away from the windows dependency now or is it still far in the future?

**Comment by ssulzer** [comment]
Hi @lojoh 
Nothing to share yet.

**Comment by SharePointX** [comment]
I have been developing a recorder application in .NET 8 for about 6 months now, using the source code of `https://github.com/microsoftgraph/microsoft-graph-comms-samples/pull/749` as a reference.
There is a limitation that it can only run on Windows, but so far I have not had any problems with the functionality.
I am eagerly waiting for the PR to be approved.
